llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp

```diff
@@ -1536,15 +1536,31 @@ RISCVTTIImpl::getArithmeticReductionCost(unsigned Opcode, VectorType *Ty,
   std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
   Type *ElementTy = Ty->getElementType();
   if (ElementTy->isIntegerTy(1)) {
+    // Example sequences:
+    //   vfirst.m a0, v0
+    //   seqz a0, a0
+    if (LT.second == MVT::v1i1)
+      return getRISCVInstructionCost(RISCV::VFIRST_M, LT.second, CostKind) +
+             getCmpSelInstrCost(Instruction::ICmp, ElementTy, ElementTy,
+                                CmpInst::ICMP_EQ, CostKind);
+
     if (ISD == ISD::AND) {
       // Example sequences:
-      //   vsetvli a0, zero, e8, mf8, ta, ma
       //   vmand.mm v8, v9, v8 ; needed every time type is split
       //   vmnot.m v8, v0
       //   vcpop.m a0, v8
       //   seqz a0, a0
-      return LT.first * getRISCVInstructionCost(RISCV::VMNAND_MM, LT.second,
-                                                CostKind) +
+
+      // Scalable VT: In nxv256i1 and larger vector elements,
+      // Fixed VT:    If getFixedSizeInBits() >= (4 * getRealMinVLen()),
+      //   the VMAND_MM instructions have started to be added.
+      InstructionCost NumOfVMAND = 0;
+      if (LT.second.isScalableVector() ||
+          LT.second.getFixedSizeInBits() == ST->getRealMinVLen())
+        NumOfVMAND = (LT.first > 2) ? (LT.first - 2) : 0;
+      return NumOfVMAND *
+                 getRISCVInstructionCost(RISCV::VMAND_MM, LT.second, CostKind) +
+             getRISCVInstructionCost(RISCV::VMNAND_MM, LT.second, CostKind) +
              getRISCVInstructionCost(RISCV::VCPOP_M, LT.second, CostKind) +
```

Comment:
Thanks for noticing the split cost! My understanding is:
  - For LMUL <= 8, there is no splitting, and the sequences are `vmnot` and `vcpop`.
  - When LMUL > 8 and split == 1, it uses `vmnand` and `vcpop`.
  - When LMUL > 8 and split > 1, the sequence becomes `vmand` * (split -1), `vmnand`, and `vcpop`.
So the cost formula would be something like
```
  cost = vcpop;
  if (LT.first == 1) cost += vmnot.
  else cost += vmnand.
  if (LT.first > 2) cost += (LT.first - 1) * vmand.
```
and this applies to both scalable VT and fixed VT.

Reply:
This indeed makes the logic much clearer. Thank you!